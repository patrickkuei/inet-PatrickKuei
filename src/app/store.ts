import articleReducer from '@inet/redux/slices/articleSlice'
import paginationReducer from '@inet/redux/slices/paginationSlice'
import searchReducer from '@inet/redux/slices/searchSlice'
import { apiSlice } from '@inet/services/apiSlice'
import { configureStore } from '@reduxjs/toolkit'

export const store = configureStore({
  reducer: {
    [apiSlice.reducerPath]: apiSlice.reducer,
    articleReducer,
    paginationReducer,
    searchReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(apiSlice.middleware),
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
