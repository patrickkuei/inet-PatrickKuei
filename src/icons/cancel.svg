<svg width="24" height="24" viewBox="0 0 24 24" fill="current"
    xmlns="http://www.w3.org/2000/svg">
    <path d="M8.46943 8.46966C8.76232 8.17677 9.23719 8.17677 9.53009 8.46966L11.9998 10.9393L14.4694 8.46967C14.7623 8.17678 15.2372 8.17678 15.5301 8.46967C15.823 8.76256 15.823 9.23744 15.5301 9.53033L13.0604 12L15.5301 14.4697C15.823 14.7626 15.823 15.2374 15.5301 15.5303C15.2372 15.8232 14.7623 15.8232 14.4694 15.5303L11.9998 13.0607L9.53009 15.5303C9.2372 15.8232 8.76233 15.8232 8.46943 15.5303C8.17654 15.2374 8.17654 14.7626 8.46943 14.4697L10.9391 12L8.46943 9.53032C8.17653 9.23743 8.17653 8.76255 8.46943 8.46966Z" fill="current"/>
    <path fill-rule="evenodd" clip-rule="evenodd" d="M7.31649 3.76876C10.404 3.42368 13.5954 3.42368 16.683 3.76876C18.5094 3.97288 19.9843 5.41153 20.1991 7.24849C20.5683 10.4054 20.5683 13.5946 20.1991 16.7515C19.9843 18.5885 18.5094 20.0271 16.683 20.2313C13.5954 20.5763 10.404 20.5763 7.31649 20.2313C5.49011 20.0271 4.01521 18.5885 3.80036 16.7515C3.43113 13.5946 3.43113 10.4054 3.80036 7.24849C4.01521 5.41153 5.49011 3.97288 7.31649 3.76876ZM16.5164 5.25948C13.5396 4.92677 10.4599 4.92677 7.4831 5.25948C6.33867 5.38738 5.42262 6.29063 5.29021 7.42274C4.93451 10.4639 4.93451 13.5361 5.29021 16.5773C5.42262 17.7094 6.33867 18.6126 7.4831 18.7405C10.4599 19.0732 13.5396 19.0732 16.5164 18.7405C17.6608 18.6126 18.5769 17.7094 18.7093 16.5773C19.065 13.5361 19.065 10.4639 18.7093 7.42274C18.5769 6.29063 17.6608 5.38738 16.5164 5.25948Z" fill="current"/>
</svg>
