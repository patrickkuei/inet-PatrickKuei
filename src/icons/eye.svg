<svg width="12" height="12" viewBox="0 0 12 12" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M4.125 6C4.125 4.96447 4.96447 4.125 6 4.125C7.03553 4.125 7.875 4.96447 7.875 6C7.875 7.03553 7.03553 7.875 6 7.875C4.96447 7.875 4.125 7.03553 4.125 6ZM6 4.875C5.37868 4.875 4.875 5.37868 4.875 6C4.875 6.62132 5.37868 7.125 6 7.125C6.62132 7.125 7.125 6.62132 7.125 6C7.125 5.37868 6.62132 4.875 6 4.875Z" fill="black"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M2.16171 5.32318C1.95215 5.62514 1.875 5.86135 1.875 6C1.875 6.13865 1.95215 6.37486 2.16171 6.67682C2.36429 6.96872 2.66539 7.28514 3.04634 7.57752C3.80989 8.16354 4.85673 8.625 6 8.625C7.14327 8.625 8.19011 8.16354 8.95366 7.57752C9.33461 7.28514 9.63571 6.96872 9.83829 6.67682C10.0478 6.37486 10.125 6.13865 10.125 6C10.125 5.86135 10.0478 5.62514 9.83829 5.32318C9.63571 5.03128 9.33461 4.71486 8.95366 4.42248C8.19011 3.83646 7.14327 3.375 6 3.375C4.85673 3.375 3.80989 3.83646 3.04634 4.42248C2.66539 4.71486 2.36429 5.03128 2.16171 5.32318ZM2.5897 3.82752C3.45483 3.16354 4.65799 2.625 6 2.625C7.34201 2.625 8.54517 3.16354 9.4103 3.82752C9.84369 4.16014 10.2016 4.53122 10.4544 4.89557C10.7003 5.24986 10.875 5.63865 10.875 6C10.875 6.36135 10.7003 6.75014 10.4544 7.10443C10.2016 7.46878 9.84369 7.83986 9.4103 8.17248C8.54517 8.83646 7.34201 9.375 6 9.375C4.65799 9.375 3.45483 8.83646 2.5897 8.17248C2.15631 7.83986 1.7984 7.46878 1.54555 7.10443C1.29969 6.75014 1.125 6.36135 1.125 6C1.125 5.63865 1.29969 5.24986 1.54555 4.89557C1.7984 4.53122 2.15631 4.16014 2.5897 3.82752Z" fill="black"/>
</svg>
