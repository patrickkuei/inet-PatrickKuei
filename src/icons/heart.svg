<svg width="12" height="12" viewBox="0 0 12 12" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M4.2 2.625C2.80957 2.625 1.625 3.66465 1.625 5.0149C1.625 5.94637 2.06117 6.73059 2.63925 7.38021C3.21533 8.0276 3.95857 8.57101 4.63049 9.02582L5.78979 9.81055C5.91676 9.89648 6.08324 9.89648 6.21021 9.81055L7.36951 9.02582C8.04143 8.57101 8.78467 8.0276 9.36075 7.38021C9.93883 6.73059 10.375 5.94637 10.375 5.0149C10.375 3.66465 9.19043 2.625 7.8 2.625C7.08326 2.625 6.45258 2.96107 6 3.39592C5.54742 2.96107 4.91674 2.625 4.2 2.625Z" fill="#E3238B"/>
</svg>
