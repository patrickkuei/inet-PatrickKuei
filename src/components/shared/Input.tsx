import { useAppDispatch } from '@inet/app/hooks'
import { CancelIcon } from '@inet/icons'
import { setSearchKeyword } from '@inet/redux/slices/searchSlice'
import clsx from 'clsx'
import { forwardRef, LegacyRef, ReactElement, useEffect, useState } from 'react'
import { useSearchParams } from 'react-router-dom'
import Button from './Button'

type Props = {
  type: string
  className?: string
  placeholder: string
  suffix: ReactElement
  onSubmit?: () => void
  onCancelClick?: () => void
}

const defaultInputClassName =
  'h-10 px-4 border border-primary-200 rounded-lg focus-within:border-primary-500 flex flex-row items-center gap-x-2 grow desktop:max-w-120 desktop:grow'

const useInputValue = (onCancelClick: (() => void) | undefined) => {
  const dispatch = useAppDispatch()
  const [searchParams] = useSearchParams()
  const [value, setValue] = useState('')
  const [shouldShowCancel, setShouldShowCancel] = useState(false)

  useEffect(() => {
    const keyword = searchParams.get('keyword')

    setValue(keyword ?? '')
    dispatch(setSearchKeyword(keyword ?? undefined))
    setShouldShowCancel(Boolean(keyword))
  }, [searchParams])

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { value: newValue } = event.target

    setValue(newValue)

    if (newValue.length > 0) {
      setShouldShowCancel(true)
    } else {
      setShouldShowCancel(false)
    }
  }

  const handleCancelClick = () => {
    setValue('')
    setShouldShowCancel(false)
    onCancelClick && onCancelClick()
  }

  return {
    value,
    shouldShowCancel,
    handleChange,
    handleCancelClick,
  }
}

const Input = (
  {
    type,
    className: customClassName = '',
    placeholder,
    suffix,
    onSubmit,
    onCancelClick,
  }: Props,
  ref: LegacyRef<HTMLInputElement> | undefined,
) => {
  const className = `${defaultInputClassName} ${customClassName}`
  const { value, shouldShowCancel, handleChange, handleCancelClick } =
    useInputValue(onCancelClick)

  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Enter') {
      onSubmit && onSubmit()
    }
  }

  return (
    <div className={className}>
      <>
        <div className="flex flex-col grow h-full py-px">
          {shouldShowCancel ? (
            <label className="text-left text-primary-500 h-2/4 text-xs">
              Search in INET
            </label>
          ) : null}
          <input
            ref={ref}
            value={value}
            type={type}
            className={clsx(
              'grow border-none outline-none placeholder:focus:text-primary-500',
              shouldShowCancel ? 'h-full' : 'h-2/4',
            )}
            placeholder={placeholder}
            onChange={handleChange}
            onKeyDown={handleKeyDown}
          />
        </div>
        {shouldShowCancel ? (
          <Button
            variant="custom"
            icon={<CancelIcon className="fill-primary-200" />}
            size="small"
            fillType="ghost"
            className="h-6 border-r border-solid rounded-none border-primary-200 pr-1.5"
            onClick={handleCancelClick}
          />
        ) : null}
      </>
      {suffix}
    </div>
  )
}

export default forwardRef<HTMLInputElement, Props>(Input)
