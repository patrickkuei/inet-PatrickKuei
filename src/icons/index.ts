import { ReactComponent as AnonymIcon } from './anonym.svg'
import { ReactComponent as BookmarkIcon } from './bookmark.svg'
import { ReactComponent as EditIcon } from './edit.svg'
import { ReactComponent as EyeIcon } from './eye.svg'
import { ReactComponent as HeartIcon } from './heart.svg'
import { ReactComponent as HistoryIcon } from './history.svg'
import { ReactComponent as LoginIcon } from './login.svg'
import { ReactComponent as LogOutIcon } from './logout.svg'
import { ReactComponent as MenuIcon } from './menu.svg'
import { ReactComponent as SearchIcon } from './search.svg'
import { ReactComponent as SettingIcon } from './setting.svg'
import { ReactComponent as SignUpIcon } from './signup.svg'
import { ReactComponent as UserIcon } from './user.svg'

export const IconMap = {
  anonym: AnonymIcon,
  bookmark: BookmarkIcon,
  edit: EditIcon,
  history: HistoryIcon,
  login: LoginIcon,
  logOut: LogOutIcon,
  menu: MenuIcon,
  search: SearchIcon,
  setting: SettingIcon,
  signup: SignUpIcon,
  user: UserIcon,
}

export type IconName = keyof typeof IconMap

export {
  AnonymIcon,
  BookmarkIcon,
  EditIcon,
  HistoryIcon,
  LoginIcon,
  LogOutIcon,
  MenuIcon,
  SearchIcon,
  SettingIcon,
  SignUpIcon,
  UserIcon,
  EyeIcon,
  HeartIcon,
}
