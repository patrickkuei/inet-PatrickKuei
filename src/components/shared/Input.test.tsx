import { store } from '@inet/app/store'
import { SearchIcon } from '@inet/icons'
import React from 'react'
import { Provider } from 'react-redux'
import { BrowserRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import Input from './Input'

it('render Input', () => {
  const ref = React.createRef<HTMLInputElement>()
  const component = renderer.create(
    <Router>
      <Provider store={store}>
        <Input
          ref={ref}
          type="text"
          placeholder="搜尋 INET"
          isGlobalSearchInput={false}
          suffix={<SearchIcon className="fill-primary-200" />}
        />
      </Provider>
    </Router>,
  )

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
