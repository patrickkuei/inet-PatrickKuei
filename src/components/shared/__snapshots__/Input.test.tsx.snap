// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Input 1`] = `
<div
  className="h-10 px-4 border border-primary-200 rounded-lg focus-within:border-primary-500 flex flex-row items-center gap-x-2 grow desktop:max-w-120 desktop:grow "
>
  <div
    className="flex flex-col grow h-full py-px"
  >
    <input
      className="grow border-none outline-none placeholder:focus:text-primary-500 h-2/4"
      onChange={[Function]}
      placeholder="搜尋 INET"
      type="text"
      value=""
    />
  </div>
</div>
`;
